## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [15]:
import pandas as pd

clients_df = pd.read_csv('Resources/client_dataset.csv')
clients_df = pd.DataFrame(clients_df)
clients_df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [16]:
# View the column names in the data
clients_df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [17]:
# Use the describe function to gather some basic statistics
clients_df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [18]:
# Use this space to do any additional research
# and familiarize yourself with the data.
clients_df.dtypes

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [19]:
# What three item categories had the most entries?
top_categories_df = clients_df.value_counts(["category"], ascending=False).head(3)
top_categories_df

category   
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [20]:
# For the category with the most entries, which subcategory had the most entries?
# Validate overall answer: bathroom supplies = 6424
# clients_df["subcategory"].value_counts()

# Find top subcategory within "consumables" category
consumables_df = clients_df.loc[(clients_df["category"] == "consumables")]
top_subcategory_df = consumables_df.value_counts(["subcategory"]).sort_values(ascending=False)
top_subcategory_df.head(1)

subcategory      
bathroom supplies    6424
Name: count, dtype: int64

In [23]:
# Which five clients had the most entries in the data?
clients_df.value_counts(["client_id"]).head(5)

client_id
33615        220
66037        211
46820        209
38378        207
24741        207
Name: count, dtype: int64

In [24]:
# Store the client ids of those top 5 clients in a list.
top_5_clients = list(clients_df['client_id'].value_counts().head(5).index)
top_5_clients


[33615, 66037, 46820, 38378, 24741]

In [27]:
# How many total units (the qty column) did the client with the most entries order?
total_units = clients_df.loc[clients_df["client_id"] == "33615"]
total_units["qty"].sum()

0

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [28]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
clients_df["line_subtotal"] = clients_df["unit_price"] * clients_df["qty"]
clients_df[["unit_price", "qty", "line_subtotal"]]

unit_price   qty  line_subtotal
0         1096.80   105      115164.00
1           24.95    21         523.95
2           13.52    39         527.28
3           36.42    29        1056.18
4          195.10    20        3902.00
...           ...   ...            ...
54634       83.13    33        2743.29
54635      206.59    47        9709.73
54636       65.66   475       31188.50
54637        1.48   112         165.76
54638        3.01  1031        3103.31

[54639 rows x 3 columns]

In [60]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def calculate_shipping(row):
    unit_weight = row["unit_weight"]
    qty = row["qty"]
    if unit_weight * qty > 50:
        shipping = (unit_weight * qty) * 7
    else:
        shipping = (unit_weight * qty) * 10
    return shipping

# Apply the function
clients_df["shipping_price"] = clients_df.apply(calculate_shipping, axis=1)

clients_df



first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_weight   qty line_number  \
0              17        2023  ...        7.50   105           1   
1              20        2023  ...        1.49    21           0   
2               4        2023  ...        1.68    39           6   
3              17        2023  ...        1.23    29           3   
4              21        2023  ...       46.43    20           1   
...           ...         ...  ...         ...   ...         ...   
54634          17        2023  ...        2.25    33           8   
54635           6        2023  ...       11.70    47           0   
54636           9        2023  ...        4.16   475           6   
54637           3        2023  ...       18.04   112           0   
54638           1        2023  ...        2.07  1031           5   

       line_subtotal  shipping_cost  shipping_price   total_price  line_cost  \
0          115164.00           75.0         5512.50  115245.93750   80159.55   
1             523.95           14.9          312.90     540.22825     331.79   
2             527.28           16.8          458.64     545.63400     323.34   
3            1056.18           12.3          356.70    1069.61775     732.95   
4            3902.00          464.3         6500.20    4409.24775    2627.70   
...              ...            ...             ...           ...        ...   
54634        2743.29           22.5          519.75    2767.87125    1725.30   
54635        9709.73          117.0         3849.30    9837.55250    8363.62   
54636       31188.50           41.6        13832.00   31233.94800   27263.85   
54637         165.76          180.4        14143.36     362.84700     317.04   
54638        3103.31           20.7        14939.19    3125.92475    1680.61   

       line_profit    line_price  
0      35086.38750  115245.93750  
1        208.43825     540.22825  
2        222.29400     545.63400  
3        336.66775    1069.61775  
4       1781.54775    4409.24775  
...            ...           ...  
54634   1042.57125    2767.87125  
54635   1473.93250    9837.55250  
54636   3970.09800   31233.94800  
54637     45.80700   

In [107]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
clients_df["line_price"] = round((clients_df["line_subtotal"] + clients_df["shipping_price"]) * 1.0925, 2)
clients_df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_weight   qty line_number  \
0              17        2023  ...        7.50   105           1   
1              20        2023  ...        1.49    21           0   
2               4        2023  ...        1.68    39           6   
3              17        2023  ...        1.23    29           3   
4              21        2023  ...       46.43    20           1   
...           ...         ...  ...         ...   ...         ...   
54634          17        2023  ...        2.25    33           8   
54635           6        2023  ...       11.70    47           0   
54636           9        2023  ...        4.16   475           6   
54637           3        2023  ...       18.04   112           0   
54638           1        2023  ...        2.07  1031           5   

       line_subtotal  shipping_cost  shipping_price   total_price  line_cost  \
0          115164.00           75.0         5512.50  115245.93750   85597.05   
1             523.95           14.9          312.90     540.22825     629.79   
2             527.28           16.8          458.64     545.63400     765.18   
3            1056.18           12.3          356.70    1069.61775    1077.35   
4            3902.00          464.3         6500.20    4409.24775    8663.60   
...              ...            ...             ...           ...        ...   
54634        2743.29           22.5          519.75    2767.87125    2222.55   
54635        9709.73          117.0         3849.30    9837.55250   12095.92   
54636       31188.50           41.6        13832.00   31233.94800   41054.25   
54637         165.76          180.4        14143.36     362.84700   14280.00   
54638        3103.31           20.7        14939.19    3125.92475   16599.10   

        line_profit  line_price  
0      46242.026250   131839.08  
1        284.468625      914.26  
2        311.937600     1077.12  
3        466.221400     1543.57  
4       2700.803500    11364.40  
...             ...         ...  
54634   1342.321200     3564.87  
54635   2717.320275    14813.24  
54636   8130.646250    49184.90  
54637   1352.713600    15632.71

In [108]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
clients_df["line_cost"] = (clients_df["unit_cost"] * clients_df["qty"]) + clients_df["shipping_price"]

clients_df



first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_weight   qty line_number  \
0              17        2023  ...        7.50   105           1   
1              20        2023  ...        1.49    21           0   
2               4        2023  ...        1.68    39           6   
3              17        2023  ...        1.23    29           3   
4              21        2023  ...       46.43    20           1   
...           ...         ...  ...         ...   ...         ...   
54634          17        2023  ...        2.25    33           8   
54635           6        2023  ...       11.70    47           0   
54636           9        2023  ...        4.16   475           6   
54637           3        2023  ...       18.04   112           0   
54638           1        2023  ...        2.07  1031           5   

       line_subtotal  shipping_cost  shipping_price   total_price  line_cost  \
0          115164.00           75.0         5512.50  115245.93750   85597.05   
1             523.95           14.9          312.90     540.22825     629.79   
2             527.28           16.8          458.64     545.63400     765.18   
3            1056.18           12.3          356.70    1069.61775    1077.35   
4            3902.00          464.3         6500.20    4409.24775    8663.60   
...              ...            ...             ...           ...        ...   
54634        2743.29           22.5          519.75    2767.87125    2222.55   
54635        9709.73          117.0         3849.30    9837.55250   12095.92   
54636       31188.50           41.6        13832.00   31233.94800   41054.25   
54637         165.76          180.4        14143.36     362.84700   14280.00   
54638        3103.31           20.7        14939.19    3125.92475   16599.10   

        line_profit  line_price  
0      46242.026250   131839.08  
1        284.468625      914.26  
2        311.937600     1077.12  
3        466.221400     1543.57  
4       2700.803500    11364.40  
...             ...         ...  
54634   1342.321200     3564.87  
54635   2717.320275    14813.24  
54636   8130.646250    49184.90  
54637   1352.713600    15632.71

In [117]:
# Create a column for the profit of each line using line cost and line price
clients_df["line_profit"] = round(clients_df["line_price"] - clients_df["line_cost"], 2)

clients_df


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_weight   qty line_number  \
0              17        2023  ...        7.50   105           1   
1              20        2023  ...        1.49    21           0   
2               4        2023  ...        1.68    39           6   
3              17        2023  ...        1.23    29           3   
4              21        2023  ...       46.43    20           1   
...           ...         ...  ...         ...   ...         ...   
54634          17        2023  ...        2.25    33           8   
54635           6        2023  ...       11.70    47           0   
54636           9        2023  ...        4.16   475           6   
54637           3        2023  ...       18.04   112           0   
54638           1        2023  ...        2.07  1031           5   

       line_subtotal  shipping_cost  shipping_price   total_price  line_cost  \
0          115164.00           75.0         5512.50  115245.93750   85597.05   
1             523.95           14.9          312.90     540.22825     629.79   
2             527.28           16.8          458.64     545.63400     765.18   
3            1056.18           12.3          356.70    1069.61775    1077.35   
4            3902.00          464.3         6500.20    4409.24775    8663.60   
...              ...            ...             ...           ...        ...   
54634        2743.29           22.5          519.75    2767.87125    2222.55   
54635        9709.73          117.0         3849.30    9837.55250   12095.92   
54636       31188.50           41.6        13832.00   31233.94800   41054.25   
54637         165.76          180.4        14143.36     362.84700   14280.00   
54638        3103.31           20.7        14939.19    3125.92475   16599.10   

       line_profit  line_price  
0         46242.03   131839.08  
1           284.47      914.26  
2           311.94     1077.12  
3           466.22     1543.57  
4          2700.80    11364.40  
...            ...         ...  
54634      1342.32     3564.87  
54635      2717.32    14813.24  
54636      8130.65    49184.90  
54637      1352.71    15632.71  
54638   

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [118]:
# Check your work using the totals above by first creating a list of the order IDs
order_id_list = [2742071, 2173913, 6128929]

# Next create a function locates the line item by order ID and then pulls the relevant line price
def find_order_price(order_id):
    order_line_df = clients_df.loc[clients_df["order_id"] == order_id, "line_price"]
    return round(order_line_df.sum(), 2)

for order_id in order_id_list:
    print(f"Order {order_id} total is: ${find_order_price(order_id)}")

find_order_price


Order 2742071 total is: $152811.89
Order 2173913 total is: $162388.71
Order 6128929 total is: $923441.25


<function __main__.find_order_price(order_id)>

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [128]:
# How much did each of the top 5 clients by quantity spend (i.e., line_price)? Check your work from Part 1 for client ids.

# Using GroupBy in order to separate the data into fields according to "client_id" values
top_buyers = clients_df.groupby(["client_id"])
top_five_spenders = top_buyers["line_price"].sum().sort_values(ascending=False).loc[top_5_clients]
top_five_spenders


client_id
33615     8377308.52
66037    10259514.79
46820     9743794.36
38378    12906550.87
24741    82268892.02
Name: line_price, dtype: float64

In [144]:
# Create a summary DataFrame showing the totals for the top 5 clients (spenders) with the following information:
# total units purchased, total shipping price, total revenue, and total profit.

# Filter the list
data = [33615, 66037, 46820, 38378, 24741]
filtered_clients = clients_df[clients_df["client_id"].isin(data)]

# Group by client ID and calculate totals
summary_df = filtered_clients.groupby("client_id").agg(
    total_units_purchased = ("qty", "sum"),
    total_shipping_price = ("shipping_price", "sum"),
    total_revenue = ("line_price", "sum"),
    total_profit = ("line_profit", "sum")
    ).reset_index()

# Display the summary DataFrame
summary_df

client_id  total_units_purchased  total_shipping_price  total_revenue  \
0      24741                 239862            5126448.37    82268892.02   
1      33615                  64313            1828984.89     8377308.52   
2      38378                  73667            3429455.40    12906550.87   
3      46820                  75768            1601448.84     9743794.36   
4      66037                  43018            1395151.85    10259514.79   

   total_profit  
0   36579992.31  
1    2201994.61  
2    3271829.89  
3    2736602.72  
4    3255031.81

In [150]:
# Format the data and rename the columns to names suitable for presentation.

presentation_df = summary_df.rename(columns={"client_id": "Client ID", 
                                             "total_units_purchased": "Total Units Purchased",
                                             "total_shipping_price": "Total Shipping Price",
                                             "total_revenue": "Total Revenue",
                                             "total_profit": "Total Profit"                                
                                             }
                                             )

# Define the money columns. 
presentation_df.info() # I see that the "money" columns (n=3) are floats. Is this what it means to define a "money column"?

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 5 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Client ID              5 non-null      int64  
 1   Total Units Purchased  5 non-null      int64  
 2   Total Shipping Price   5 non-null      float64
 3   Total Revenue          5 non-null      float64
 4   Total Profit           5 non-null      float64
dtypes: float64(3), int64(2)
memory usage: 328.0 bytes


In [160]:
# Define a function named currency_format_millions that converts a dollar amount to millions.
def currency_format_millions(x):
    amount_millions = x / 1000000
    return amount_millions

# Apply the currency_format_millions function to only the money columns. 
presentation_df[["Amount1 in Millions", 
                 "Amount2 in Millions", 
                 "Amount3 in Millions"]] = presentation_df[["Total Shipping Price", 
                                                            "Total Revenue", 
                                                            "Total Profit"]].apply(currency_format_millions)

# Rename the columns to reflect the change in the money format.
presentation_df = presentation_df.rename(columns={"Amount1 in Millions": "Total Shipping Price ($ Millions)", 
                                                  "Amount2 in Millions": "Total Revenue ($ Millions)", 
                                                  "Amount3 in Millions": "Total Profit ($ Millions)"})

# drop duplicates
presentation_df = presentation_df.loc[:, ~presentation_df.columns.duplicated()]

# Display the dataframe
presentation_df


Client ID  Total Units Purchased  Total Shipping Price  Total Revenue  \
0      24741                 239862            5126448.37    82268892.02   
1      33615                  64313            1828984.89     8377308.52   
2      38378                  73667            3429455.40    12906550.87   
3      46820                  75768            1601448.84     9743794.36   
4      66037                  43018            1395151.85    10259514.79   

   Total Profit  Total Shipping Price ($ Millions)  \
0   36579992.31                           5.126448   
1    2201994.61                           1.828985   
2    3271829.89                           3.429455   
3    2736602.72                           1.601449   
4    3255031.81                           1.395152   

   Total Revenue ($ Millions)  Total Profit ($ Millions)  
0                   82.268892                  36.579992  
1                    8.377309                   2.201995  
2                   12.906551                   3.271830  
3                    9.743794                   2.736603  
4                   10.259515                   3.255032

In [162]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.

final_df = presentation_df.sort_values("Total Profit ($ Millions)", ascending=False)

final_df

Client ID  Total Units Purchased  Total Shipping Price  Total Revenue  \
0      24741                 239862            5126448.37    82268892.02   
2      38378                  73667            3429455.40    12906550.87   
4      66037                  43018            1395151.85    10259514.79   
3      46820                  75768            1601448.84     9743794.36   
1      33615                  64313            1828984.89     8377308.52   

   Total Profit  Total Shipping Price ($ Millions)  \
0   36579992.31                           5.126448   
2    3271829.89                           3.429455   
4    3255031.81                           1.395152   
3    2736602.72                           1.601449   
1    2201994.61                           1.828985   

   Total Revenue ($ Millions)  Total Profit ($ Millions)  
0                   82.268892                  36.579992  
2                   12.906551                   3.271830  
4                   10.259515                   3.255032  
3                    9.743794                   2.736603  
1                    8.377309                   2.201995

# Module 4 Summary

From 2022 - 2023, the company received 54,639 orders. During that time, client 24741 was by far the most active in multiple categories that include: generating total profit ($36.58M); generating total revenue ($82.27M); and total units purchased (239,862).  